text/3691-trusted-publishing-cratesio.md

```diff
@@ -0,0 +1,371 @@
+    - For the workflow to be authorized to do this, [it needs `id-token: write` permissions declared in the workflow YAML](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#adding-permissions-settings).
+3. After authenticating the workflow's request for an ID token, the GitHub Actions OIDC provider returns a cryptographically signed ID token (signed by its private key) .
+4. The release workflow, sends the signed ID token to crates.io to exchange for an access token.
+    - The ID token is not used to authorize any action on crates.io, it's only used to authenticate.
+5. Crates.io, using the public keys of the GitHub Action OIDC Provider, validates the signature of the provided ID token that has been signed.
+6. Once validated, crates.io generates a temporary access token and returns it to the Github Actions release workflow.
+7. The release workflow builds the artifact to be published (i.e. [the `.crate` file](https://doc.rust-lang.org/cargo/commands/cargo-package.html))
+8. The release workflow, using the access token retrieved from step 6, authenticates to crates.io and publishes the version of the crate.
+9. As a good steward, once the crate has successfully been published, the workflow job will request the revocation of the access token from crates.io
+    - An access token's lifetime might outlive the actual time needed to perform the action. It's good practice to revoke it once done to further limit any risks of accidental exposure. 
+
+> [!NOTE]
+> Steps 7 and 8 are abstracted in the `cargo publish` command. The access token would be passed via the `CARGO_REGISTRY_TOKEN` environment variable.
+ 
+</details>
+
+### Trusted Publisher Configuration on crates.io
+[trusted-publisher-configuration-cratesio]: #trusted-publisher-configuration-cratesio
+
+Prior to publishing a crate through a GitHub Actions workflow, the crate author will need to create a _Trusted Publisher Configuration_ via the crates.io web UI. This is a one-time configuration **required per crate**.
+
+> [!Note]
+> Initially, a _Trusted Publisher Configuration_ can only be created after an initial manual publishing of a crate.
+> 
+> See [Future Possibilities](#future-possibilities) for future plans to alleviate this pain.
+
+Without a corresponding _Trusted Publisher Configuration_ any attempts from a GitHub Action workflow to exchange an ID token for an access token from crates.io will be unauthorized and fail. 
+
+A _Trusted Publisher Configuration_ will vary on requirements depending on the trusted publisher being used (e.g. GitHub Actions, GitLab Pipelines, and so on). CI/CD providers will vary in the claims they make within their ID token. 
+
+For GitHub Actions, the crate author will need to provide:
+
+1. (required) The owning GitHub username or organization
+2. (required) The repository name
+3. (required) The workflow file name (must be located in `.github/workflows/`)
```

Comment:
It's not strictly required, but I would strongly recommend requiring it: making this claim optional would mean that any workflow in the source repository could generate an OIDC token with the ability publish. Requiring it to be scoped to a single workflow significantly reduces the opportunity for compromise here.

Additionally, the use case of wanting to publish from multiple workflows with a single configuration is not a strong one, I think.  Usually there is just a single workflow which can publish, and if a user _really_ needs to publish from multiple workflows, they can always configure multiple publishers with varying workflow filenames instead.

Reply:
The threat would be third-party code that is necessary for the workflow to function being compromised to steal OIDC credentials (or API tokens, for that matter).

Consider two separate workflows: one that does testing & linting, another that handles publishing on release. The former may introduce a lot of dependencies that aren't required for publishing, but are required for testing/linting. If this test/lint workflow also has the ability to generate OIDC credentials that can be used for publishing, this unnecessarily increases the surface area that the OIDC token is exposed to.

(Note that for GitHub as an IdP specifically, this may be sort of a moot point, because providing the token to the workflow requires an explicit configuration in the workflow with `id-token: write`, however this is fairly easy to over-configure.)

Reply:
And specifically, actions used inside of workflows *not named `release.yml`* being compromised is a lot easier to contemplate. This functionally reduces the scope of your audit to just the `release` workflow.

Reply:
I've added a note about considering supply chain compromises within the release workflow. We'll keep the workflow file name as required.